### PR TITLE
Modulesync 5.1.0

### PR DESCRIPTION
**Pull request by smortex**

### CODE REVIEW

File: spec/functions/generate_statement_spec.rb [hunk 1]
```diff
@@ -157,7 +159,7 @@
         {
           'type' => 'syslog',
           'options' => [
-            { 'flags' => ['no-multi-line', 'no-parse'] },
```
Comment (faxm0dem) [inline spec/functions/generate_statement_spec.rb]: rubocop-- : I think this was way more readable before :-1:

File: spec/defines/module_spec.rb [hunk 2]
```diff
@@ -11,7 +13,7 @@
   end
   let :facts do
     default_facts.merge(
-      { osfamily: 'Debian', os: { family: 'Debian', name: 'Ubuntu', release: { full: '14.4', major: '14.4' } } },
+      { osfamily: 'Debian', os: { family: 'Debian', name: 'Ubuntu', release: { full: '14.4', major: '14.4' } } }
```
Comment (faxm0dem) [inline spec/defines/module_spec.rb]: :man_facepalming:

File: lib/puppet/parser/functions/statement.rb [hunk 3]
```diff
@@ -38,6 +38,7 @@ def build; end
   # like the file name in a file source
   class SimpleParameter < Parameter
     def initialize(value = '')
+      super()
```
Comment (faxm0dem) [inline lib/puppet/parser/functions/statement.rb:41]: out of curiosity, what's the purpose of this change ? :superhero:

File: spec/unit/facter/syslog_ng_version_spec.rb [hunk 4]
```diff
@@ -37,20 +37,26 @@
         'Enable-Linux-Caps' => 'off',
       }
       expected_syslog_ng_version = '3.7.1'
+
```
Comment (faxm0dem) [inline spec/unit/facter/syslog_ng_version_spec.rb:40]: Could you please explain the change ?
The new form doesn't make sense to me : as I understand it, the `before` fixes that `which syslog-ng` returns `/usr/bin/syslog-ng`, and the test tests what we just fixed :question:

**Review by faxm0dem — COMMENTED** [review]
thanks! Please find my comments in-line

File: lib/puppet/parser/functions/statement.rb (hunk 3 above)
Comment (smortex) [reply lib/puppet/parser/functions/statement.rb:41]: The constructor in the base class (10 lines above) is not called.  As a result, without this:

```ruby
x = SimpleParameter.new
p x.instance_variable_get('@params') #=> nil
p x.instance_variable_get('@value')  #=> ""
```

File: spec/defines/module_spec.rb (hunk 2 above)
Comment (smortex) [reply spec/defines/module_spec.rb]: We can probably tune this in rubocop config.

File: spec/functions/generate_statement_spec.rb (hunk 1 above)
Comment (smortex) [reply spec/functions/generate_statement_spec.rb]: I trend to agree :-)  This is tunable in rubocop.  Maybe it is worth providing our own rubocop config instead of relying on the one from voxpupuli; or add overrides to rubocop.yml and use modulesync to use this for the in2p3 modules.

File: spec/unit/facter/syslog_ng_version_spec.rb (hunk 4 above)
Comment (smortex) [reply spec/unit/facter/syslog_ng_version_spec.rb:40]: Both code are equivalent, but RSpec is a bit more explicit and split the 2 things that mocha do in 2 operations:
1. `allow` describe what *can* happen and stubs it;
2. `expect` ensure it happened, raising an exception if it did not.

Mocha with a single `expects` before the code being trained setup the stubs and register code to check it was called at the end of the test.

File: spec/functions/generate_statement_spec.rb (hunk 1 above)
Comment (smortex) [reply spec/functions/generate_statement_spec.rb]: Opened #24 regarding this.

File: spec/defines/module_spec.rb (hunk 2 above)
Comment (smortex) [reply spec/defines/module_spec.rb]: Opened #25 regarding this.

File: spec/functions/generate_statement_spec.rb (hunk 1 above)
Comment (smortex) [reply spec/functions/generate_statement_spec.rb]: https://github.com/ccin2p3/modulesync_config/pull/2 address this. I amended the commit to sync with the result of the merge.

File: spec/defines/module_spec.rb (hunk 2 above)
Comment (smortex) [reply spec/defines/module_spec.rb]: https://github.com/ccin2p3/modulesync_config/pull/2 address this. I amended the commit to sync with the result of the merge.

File: manifests/repo.pp [hunk 5]
```diff
@@ -35,7 +35,7 @@
           },
         }
 
-        Class['apt::update'] -> Package <| tag == 'syslog_ng' |>
```
Comment (faxm0dem) [inline manifests/repo.pp:38]: Could you please explain this change ?

File: spec/unit/facter/syslog_ng_version_spec.rb (hunk 4 above)
Comment (faxm0dem) [reply spec/unit/facter/syslog_ng_version_spec.rb:40]: :pray:

File: lib/puppet/parser/functions/statement.rb (hunk 3 above)
Comment (faxm0dem) [reply lib/puppet/parser/functions/statement.rb:41]: why did it work before ?

Comment (smortex) [reply lib/puppet/parser/functions/statement.rb:41]: `@params` seems not to be used, at least not in the 2 sub-classes of `Parameter` in this module :shrug:.

File: manifests/repo.pp (hunk 5 above)
Comment (smortex) [reply manifests/repo.pp:38]: This is related to this failure https://github.com/ccin2p3/puppet-syslog_ng/runs/5365004194?check_suite_focus=true introduced by https://github.com/puppetlabs/puppetlabs-apt/commit/966ca80c56af071e53179d6d04cdd21f6523121f

```
        dependency cycles found: (Exec[apt_update] => Class[Apt::Update] => Package[apt-transport-https] => Class[Apt::Update] => Exec[apt_update])
```

1. the class `syslog_ng` include `syslog_ng::repo`
2. `syslog_ng::repo` declare a `apt::source`
3. `apt::source` use create_resource to declare a package `apt-transport-https` if it does not exist.

So if `apt-transport-https` is missing before the module is declared, the `Package['apt-transport-https']` resource will have tags `syslog_ng, syslog_ng::repo, apt::source` (scopes of the calling contexts).

Then the dependencies come in place:
```
Package['apt-transport-https'] -> Class['apt::update']
Class['apt::update'] -> Package <| tag == 'syslog_ng' |>
```

Because `Package['apt-transport-https']` is tagged `syslog_ng` this is a circular dependency.
